src/runtime/finalizer.cs

```diff
@@ -178,12 +123,18 @@ private void DisposeAll()
                     try
                     {
                         obj.Dispose();
-                        Runtime.CheckExceptionOccurred();
```

Comment:
Remove this can save the overhead, but it may ignore some exceptions.
When reference count become 0, tp_dealloc may call, some internal exceptions may cause in there.
A better way is check the reference count is 1 or not, if it's 1, check the exception after decref the obj, of course these should implement inside Dispose.

Reply:
It is hard to find documentation about this. According to [this SO answer](https://stackoverflow.com/questions/52468416/who-should-call-pyerr-fetch/52469036#52469036) `Py_DecRef` should preserve exception state.

Reply:
I encountered some scenes in the past: `tp_dealloc` raised an exception but no one catch it until back to the script then some other calls check and raise, if this happened, you can't locate the exception precisely at the short time. Thus I simply add this.
According to other C source within python, you can assume that `tp_dealloc` wouldn't raise any exceptions, so it's acceptable to remove it.
Nevertheless, `check the reference count is 1 or not, if it's 1, check the exception after decref the obj, of course these should implement inside Dispose` in the debug mode still can be a option to make it expose the bugs easily.

Reply:
@amos402 problem is, our `Dispose` might be called during someone else's `tp_dealloc`, and raising that exception in C# will cause their `tp_dealloc` to fail with a different exception, thus violating the requirement for `tp_dealloc`  to preserve exceptions.

Reply:
Then it should be `Py_FatalError` instead an exception throw in `Dispose`, of course since it would make some performance impact keep it in debug mode.

Reply:
@amos402 I am still unconvinced by this. Problem is the Python documentation is very lacking on such important details.

Here's the scenario I have in mind:

```python
a = SomeClass()
try:
  foo()
finally:
  a = None # <- this calls __del__ and might eventually call our Dispose
```

I am unsure about Python behavior here, but if `a = None` is executed with an error flag set (because `foo()` threw an exception), then doing `Runtime.CheckExceptionOccurred()` in `Dispose` will cause the exception to be reraised inside `finally` block, which would be totally wrong.

Now that scenario might not behave as I described, but I am unsure if that scenario is impossible.

The reason I don't want to put this behind an `#if DEBUG` is that I don't want switching to Debug build to cause unrelated major change in behavior.

Reply:
Not calling `Runtime.CheckExceptionOccurred()`, use `Assert` or `Py_FatalError` instead, the purpose the for exposing the error as soon as possible and make that locatable.

https://github.com/python/cpython/blob/3d53d8756f0403eec6a4e12f183103d651bed6c5/Python/ceval.c#L3691-L3699
Like this, it can help developer find out the problem quickly. Not the same case, I just give an example.

Reraise an exception wouldn't be ignore by Python3, it will throw an exception chain.
Try this:
```python
try:
    rasie Exception(1)
finally:
    raise Exception(2)
```
Following the C sources, not only depending the document should be helpful.

Reply:
@amos402 as I said, I was not implying this exact example breaks. Forget it. I found a relevant piece in [the documentation](https://docs.python.org/3/extending/newtypes.html#finalization-and-de-allocation):

> One important requirement of the deallocator function is that it leaves any pending exceptions alone. This is important since deallocators are frequently called as the interpreter unwinds the Python stack; when the stack is unwound due to an exception (rather than normal returns), nothing is done to protect the deallocators from seeing that an exception has already been set. Any actions which a deallocator performs which may cause additional Python code to be executed may detect that an exception has been set. This can lead to misleading errors from the interpreter.

Reply:
`Any actions which a deallocator performs which may cause additional Python code to be executed may detect that an exception has been set. This can lead to misleading errors from the interpreter.`
That's what I said before. That's why you shouldn't ignore the exception, especially on debug build.

Reply:
@amos402 this is exactly why we must ignore the exception! Our `Dispose` can be called during Python stack unwinding as described in this quote, and it should totally ignore the exception, that caused the unwinding, otherwise it will cause double fault.

Reply:
https://gist.github.com/amos402/5eb5db264bf63414db395fdc4fd178e1
Try it on both Debug and Release of CPython.

Reply:
@amos402 I am still unconvinced, but I feel like this change might be important for general stability, so I've added the check to debug builds, as there's little harm in doing so.